sklearn/utils/estimator_checks.py

```diff
@@ -2565,6 +2567,60 @@ def check_classifiers_regression_target(name, estimator_orig):
         assert_raises_regex(ValueError, msg, e.fit, X, y)
 
 
+@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+def check_estimator_sparse_dense(name, estimator_orig):
+    rng = np.random.RandomState(52)
+    estimator = clone(estimator_orig)
+    estimator_sp = clone(estimator_orig)
+    tags = _safe_tags(estimator_orig)
+    centers = 2 if tags["binary_only"] else None
+    X, y = make_blobs(random_state=rng, cluster_std=0.5, centers=centers)
+    # for put some points to zero to have a little bit of sparsity
+    # TODO: update this with enforce_estimator_tags_X when #14705 is merged
+    if tags["requires_positive_X"]:
+        X -= np.min(X) - 1.
+    X_csr = sparse.csr_matrix(X)
+    if tags["multioutput"] or tags["multioutput_only"]:
+        y = y[:, np.newaxis]
+
+    for sparse_format in ['csr', 'csc', 'dok', 'lil', 'coo', 'dia', 'bsr']:
+        X_sp = X_csr.asformat(sparse_format)
+        if name in ['Scaler', 'StandardScaler']:
+            estimator.set_params(with_mean=False)
+            estimator_sp.set_params(with_mean=False)
+        if name in ['RobustScaler']:
+            estimator.set_params(with_centering=False)
+            estimator_sp.set_params(with_centering=False)
+
+        set_random_state(estimator)
+        set_random_state(estimator_sp)
+        X_converted = pairwise_estimator_convert_X(X, estimator)
+        X_sp_converted = pairwise_estimator_convert_X(X_sp, estimator)
+        try:
+            estimator_sp.fit(X_sp_converted, y)
+            estimator.fit(X_converted, y)
+            if hasattr(estimator, "predict"):
+                pred = estimator.predict(X_converted)
+                pred_sp = estimator_sp.predict(X_sp_converted)
+                if not (isinstance(estimator, BaseSGD) or  # SGD is too random
+                        getattr(estimator, 'kernel', None) == 'precomputed' or
+                        getattr(estimator, 'metric', None) == 'precomputed'):
+                    assert_allclose(pred, pred_sp)
+
+                assert pred.shape == pred_sp.shape
+            if hasattr(estimator, 'predict_proba'):
+                probs = estimator.predict_proba(X_converted)
+                probs_sp = estimator.predict_proba(X_sp_converted)
+                assert probs_sp.shape == (X.shape[0], len(np.unique(y)))
+                assert_allclose(probs, probs_sp)
+        except (ValueError, TypeError):
```

Comment:
In the end I decided to keep two different tests, but to remove the part where we check the error message if sparse inputs is not supported here, since it's the part already tested by `check_estimator_sparse_data`
What do you think ?

Reply:
This is a problem because `ValueError` or `TypeError` could be raised by the code in the checks, after you call `fit`. For instance when you call `predict_proba` in the check.

So please instead do something like:

```

        try:
            estimator_sp.fit(X_sp_converted, y)
            estimator.fit(X_converted, y)
         except (ValueError, TypeError):
            # This is the case where estimators don't support sparse inputs.
            # We don't want to have this case interfere  with our test of equality:
            # it is already tested by check_estimator_sparse_data
            continue
        # put the rest of the code specific to check_estimator_sparse_dense
        # here
        ...
```

Reply:
I agree, done, I used the try/except/else statement, so that only estimators that raised no exception (so fitted in the "try"), will be asked to `predict`.